File: projetos-fenix-para-estudo/t2ti-erp-fenix-cadastros-completo/fontes/frontend/provider/lib/src/service/cadastros/produto_subgrupo_service.dart
/*
Title: T2Ti ERP Fenix                                                                
Description: Service relacionado à tabela [PRODUTO_SUBGRUPO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';
import 'package:http/http.dart' show Client;

import 'package:fenix/src/service/service_base.dart';
import 'package:fenix/src/model/filtro.dart';
import 'package:fenix/src/model/model.dart';

/// classe responsável por requisições ao servidor REST
class ProdutoSubgrupoService extends ServiceBase {
  var clienteHTTP = Client();

  Future<List<ProdutoSubgrupo>> consultarLista({Filtro filtro}) async {
    var listaProdutoSubgrupo = List<ProdutoSubgrupo>();

    tratarFiltro(filtro, '/produto-subgrupo/');
    final response = await clienteHTTP.get(url);

    if (response.statusCode == 200) {
      if (response.headers["content-type"].contains("html")) {
        tratarRetornoErro(response.body, response.headers);
        return null;
      } else {
        var parsed = json.decode(response.body) as List<dynamic>;
        for (var produtoSubgrupo in parsed) {
          listaProdutoSubgrupo.add(ProdutoSubgrupo.fromJson(produtoSubgrupo));
        }
        return listaProdutoSubgrupo;
      }
    } else {
      tratarRetornoErro(response.body, response.headers);
      return null;
    }
  }

  Future<ProdutoSubgrupo> consultarObjeto(int id) async {
    final response = await clienteHTTP.get('$endpoint/produto-subgrupo/$id');

    if (response.statusCode == 200) {
      if (response.headers["content-type"].contains("html")) {
        tratarRetornoErro(response.body, response.headers);
        return null;
      } else {
        var produtoSubgrupoJson = json.decode(response.body);
        return ProdutoSubgrupo.fromJson(produtoSubgrupoJson);
      }
    } else {
      tratarRetornoErro(response.body, response.headers);
      return null;
    }
  }

  Future<ProdutoSubgrupo> inserir(ProdutoSubgrupo produtoSubgrupo) async {
    final response = await clienteHTTP.post(
      '$endpoint/produto-subgrupo',
      headers: {"content-type": "application/json"},
      body: produtoSubgrupo.objetoEncodeJson(produtoSubgrupo),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      if (response.headers["content-type"].contains("html")) {
        tratarRetornoErro(response.body, response.headers);
        return null;
      } else {
        var produtoSubgrupoJson = json.decode(response.body);
        return ProdutoSubgrupo.fromJson(produtoSubgrupoJson);
      }
    } else {
      tratarRetornoErro(response.body, response.headers);
      return null;
    }
  }

  Future<ProdutoSubgrupo> alterar(ProdutoSubgrupo produtoSubgrupo) async {
    var id = produtoSubgrupo.id;
    final response = await clienteHTTP.put(
      '$endpoint/produto-subgrupo/$id',
      headers: {"content-type": "application/json"},
      body: produtoSubgrupo.objetoEncodeJson(produtoSubgrupo),
    );

    if (response.statusCode == 200) {
      if (response.headers["content-type"].contains("html")) {
        tratarRetornoErro(response.body, response.headers);
        return null;
      } else {
        var produtoSubgrupoJson = json.decode(response.body);
        return ProdutoSubgrupo.fromJson(produtoSubgrupoJson);
      }
    } else {
      tratarRetornoErro(response.body, response.headers);
      return null;
    }
  }

  Future<bool> excluir(int id) async {
    final response = await clienteHTTP.delete(
      '$endpoint/produto-subgrupo/$id',
      headers: {"content-type": "application/json"},
    );

    if (response.statusCode == 200) {
      return true;
    } else {
      tratarRetornoErro(response.body, response.headers);
      return null;
    }
  }
}

File: projetos-fenix-para-estudo/t2ti-erp-fenix-cadastros-completo/fontes/backend/delphi/dmvc/Service/Cadastros/ProdutoSubgrupoService.pas
{*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Service relacionado à tabela [PRODUTO_SUBGRUPO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************}
unit ProdutoSubgrupoService;

interface

uses
  ProdutoSubgrupo, ProdutoGrupo, 
  System.SysUtils, System.Generics.Collections, ServiceBase, MVCFramework.DataSet.Utils;

type

  TProdutoSubgrupoService = class(TServiceBase)
  private
    class procedure AnexarObjetosVinculados(AListaProdutoSubgrupo: TObjectList<TProdutoSubgrupo>); overload;
    class procedure AnexarObjetosVinculados(AProdutoSubgrupo: TProdutoSubgrupo); overload;
  public
    class function ConsultarLista: TObjectList<TProdutoSubgrupo>;
    class function ConsultarListaFiltroValor(ACampo: string; AValor: string): TObjectList<TProdutoSubgrupo>;
    class function ConsultarObjeto(AId: Integer): TProdutoSubgrupo;
    class procedure Inserir(AProdutoSubgrupo: TProdutoSubgrupo);
    class function Alterar(AProdutoSubgrupo: TProdutoSubgrupo): Integer;
    class function Excluir(AProdutoSubgrupo: TProdutoSubgrupo): Integer;
  end;

var
  sql: string;


implementation

{ TProdutoSubgrupoService }

class procedure TProdutoSubgrupoService.AnexarObjetosVinculados(AProdutoSubgrupo: TProdutoSubgrupo);
begin
  // ProdutoGrupo
  sql := 'SELECT * FROM PRODUTO_GRUPO WHERE ID = ' + AProdutoSubgrupo.IdProdutoGrupo.ToString;
  AProdutoSubgrupo.ProdutoGrupo := GetQuery(sql).AsObject<TProdutoGrupo>;

end;

class procedure TProdutoSubgrupoService.AnexarObjetosVinculados(AListaProdutoSubgrupo: TObjectList<TProdutoSubgrupo>);
var
  ProdutoSubgrupo: TProdutoSubgrupo;
begin
  for ProdutoSubgrupo in AListaProdutoSubgrupo do
  begin
    AnexarObjetosVinculados(ProdutoSubgrupo);
  end;
end;

class function TProdutoSubgrupoService.ConsultarLista: TObjectList<TProdutoSubgrupo>;
begin
  sql := 'SELECT * FROM PRODUTO_SUBGRUPO ORDER BY ID';
  try
    Result := GetQuery(sql).AsObjectList<TProdutoSubgrupo>;
    AnexarObjetosVinculados(Result);
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TProdutoSubgrupoService.ConsultarListaFiltroValor(ACampo, AValor: string): TObjectList<TProdutoSubgrupo>;
begin
  sql := 'SELECT * FROM PRODUTO_SUBGRUPO where ' + ACampo + ' like "%' + AValor + '%"';
  try
    Result := GetQuery(sql).AsObjectList<TProdutoSubgrupo>;
    AnexarObjetosVinculados(Result);
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TProdutoSubgrupoService.ConsultarObjeto(AId: Integer): TProdutoSubgrupo;
begin
  sql := 'SELECT * FROM PRODUTO_SUBGRUPO WHERE ID = ' + IntToStr(AId);
  try
    GetQuery(sql);
    if not Query.Eof then
    begin
      Result := Query.AsObject<TProdutoSubgrupo>;
      AnexarObjetosVinculados(Result);
    end
    else
      Result := nil;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class procedure TProdutoSubgrupoService.Inserir(AProdutoSubgrupo: TProdutoSubgrupo);
begin
  AProdutoSubgrupo.ValidarInsercao;
  AProdutoSubgrupo.Id := InserirBase(AProdutoSubgrupo, 'PRODUTO_SUBGRUPO');
  
end;

class function TProdutoSubgrupoService.Alterar(AProdutoSubgrupo: TProdutoSubgrupo): Integer;
begin
  AProdutoSubgrupo.ValidarAlteracao;
  Result := AlterarBase(AProdutoSubgrupo, 'PRODUTO_SUBGRUPO');
  
  
end;


class function TProdutoSubgrupoService.Excluir(AProdutoSubgrupo: TProdutoSubgrupo): Integer;
begin
  AProdutoSubgrupo.ValidarExclusao;
  
  Result := ExcluirBase(AProdutoSubgrupo.Id, 'PRODUTO_SUBGRUPO');
end;


end.
